### PR TITLE
feat: extensible commands

### PR DESCRIPTION
**Pull request by lengau**
This creates a version of commands that can be modified by the application by registering parser fillers, prologues and epilogues. Additionally, it makes the lifecycle commands versions of these.

This is needed to support charmcraft's craft-application changeover, but is kept in draft until I can test it with charmcraft.

- [x] Have you followed the guidelines for contributing?
- [x] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?
- [x] Have you successfully run `tox`?

-----

### CODE REVIEW

File: craft_application/commands/base.py [hunk 1]
```diff
@@ -74,3 +94,83 @@ def get_managed_cmd(
...
+    """A command that allows applications to register modifications."""
+
+    _parse_callback: ParserCallback | None
+    _prologue: RunCallback | None
+    _epilogue: RunCallback | None
+
+    @property
+    def services(self) -> service_factory.ServiceFactory:
+        """Services available to this command."""
+        return self._services  # pragma: no-cover
```
[the remark this answers is not in the document]
Comment (lengau) [reply craft_application/commands/base.py:109]: Here's the sample from my Snapcraft craft-application work:

https://github.com/snapcore/snapcraft/pull/4449/commits/9d4809df16a2733bd064f8564d0cae219edc54c4#diff-b150581a4616ec37d400371c1b2ef27849f7bc5b1b7e7e691e1ecce52d5e657cR90-R110

If the parsed arguments include `--enable-manifest`, the prologue tells the lifecycle service to enable manifests. (This uses a snapcraft-specific additional method on the LifecycleService.)

**Review by mr-cal — APPROVED** [review]
LGTM but I would also like to see how it is used in the wild.

File: craft_application/commands/base.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (tigarmo) [reply craft_application/commands/base.py]: thanks. So the idea is to be able to extend the base commands' behaviors without subclassing them, right?

**Review by tigarmo — APPROVED** [review]

File: craft_application/commands/base.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (lengau) [reply craft_application/commands/base.py]: Yep! It allows for a much more restricted form of extending these behaviours than having to subclass each of them. Lifecycle commands are probably the primary use case here. If an application needs to add a parameter to all lifecycle commands, this seems a whole lot cleaner than subclassing each of the lifecycle commands to add the parameter and change the behaviour (with then also the temptation to use multiple inheritance...)

I don't currently see a need for this outside of the lifecycle commands, though I suppose there could potentially be something for store commands.

cc @mr-cal your example is above :-)

[the remark this answers is not in the document]
Comment (mr-cal) [reply craft_application/commands/base.py]: Thanks!

**Pull request merged**